### PR TITLE
Fix MEND archive loading on newer PyTorch

### PR DESCRIPTION
**Pull request by xwh12345-user**

### CODE REVIEW

**Comment by xwh12345-user** [comment]
This PR fixes MEND training/editing for Qwen2.5 models, especially when the base model is loaded with model parallelism.

  What Changed

  - Fixed MEND edited-parameter application for Qwen/Qwen2.5.
  - Added a differentiable parameter replacement wrapper for MEND edited models.
  - For Qwen models, MEND no longer relies on the previous higher/monkeypatch path to apply p + update.
  - During edited-model forward, the wrapper temporarily swaps the edited tensors into the base model parameters, runs the
    forward pass, then restores the original parameters.

  - This keeps the post-edit loss connected to MEND’s outer parameters, so MEND’s small editor network can actually receive gradients.

  This fixes the previous issue where MEND generated updates, but Qwen2.5 forward did not actually use the updated
  parameters. As a result, training appeared to run but MEND parameters were not updated.

  Also retained the MEND archive loading fix:

  - Use torch.load(..., weights_only=False) when loading MEND archives, because newer PyTorch versions otherwise reject
    archives containing non-tensor objects such as model configs.

  Why This Was Needed

  Before this fix, Qwen2.5 MEND training showed the following symptoms:

  OUTER_PARAMS 21
  GRAD_NONE 21
  GRAD_NONZERO 0
  grad_train: 0.00000

  Validation loss stayed unchanged across training steps, optimizer state remained empty, and edited logits were identical
  to base logits:

  LOGIT_MAX_ABS_DIFF 0.0
  LOGIT_MEAN_ABS_DIFF 0.0

  The root cause was that the old functional/monkeypatch edited model path did not reliably apply updated parameters to
  Qwen2.5 forward, especially with device_map / model-parallel loading.

  Validation

  Tested with:

  - Model: /disk4/xuweihong/models/Qwen2.5-7B-Instruct
  - GPUs for multi-GPU validation: CUDA_VISIBLE_DEVICES=1,4,5
  - Training method: MEND
  - Synthetic ZsRE-style smoke data

  Multi-GPU gradient diagnostic after the fix:

  OUTER_PARAMS 21
  GRAD_NONE 0
  GRAD_NONZERO 13
  GRAD_NORM_SUM 21457.857...

  Multi-GPU training + edit validation:

  TRAIN_HF_DEVICE_MAP_DEVICES ['0', '1', '2']
  EDIT_HF_DEVICE_MAP_DEVICES ['0', '1', '2']
  RESULT MEND TRAIN+EDIT 3GPU OK Qwen2ForCausalLM same_object False

  Training metrics now show non-zero gradients and changing validation loss:

  grad_train: 49440.28125
  grad_train: 38494.62109
  grad_train: 17304.86719

  loss/edit_val: 81.40489
  loss/edit_val: 80.89652
  loss/edit_val: 61.57682

  Saved archive inspection confirmed real optimizer/parameter updates:

  archive_step 6
  opt_state_entries 21
  edit_lrs changed from initial values
  MEND MLP parameters no longer all zero

  Single-GPU Qwen2.5 validation also passed with model_parallel=False:

  VISIBLE 1
  MODEL_PARALLEL False
  EDITED_MODEL_TYPE EditedParameterModel
  LOGIT_MAX_ABS_DIFF 86.0
  LOGIT_MEAN_ABS_DIFF 3.3050
  OUTER_PARAMS 21
  GRAD_NONE 0
  GRAD_NONZERO 13

  Single-GPU non-model-parallel training also completed successfully:

  MODEL_PARALLEL False
  grad_train: 49440.28125
  grad_train: 32662.22852
  grad_train: 18371.33789
  RESULT SINGLE NONMP TRAIN OK

**Comment by xwh12345-user** [comment]
"""Quick single/multi-GPU verification for MEND child-level monkeypatch."""

import argparse
import os
import subprocess
import sys

import torch
from transformers import AutoTokenizer

REPO = "/disk4/xuweihong/workspace/EasyEdit"
MODEL = "/disk4/xuweihong/models/Qwen2.5-7B-Instruct"
PYTHON = "/disk4/xuweihong/envs/miniconda3/envs/EasyEdit/bin/python"
INNER_PARAMS = [
    "model.layers.25.mlp.gate_proj.weight",
    "model.layers.25.mlp.up_proj.weight",
    "model.layers.25.mlp.down_proj.weight",
    "model.layers.26.mlp.gate_proj.weight",
    "model.layers.26.mlp.up_proj.weight",
    "model.layers.26.mlp.down_proj.weight",
    "model.layers.27.mlp.gate_proj.weight",
    "model.layers.27.mlp.up_proj.weight",
    "model.layers.27.mlp.down_proj.weight",
]


def configure(mode):
    from easyeditor import MENDTrainingHparams

    hparams = MENDTrainingHparams.from_hparams(
        f"{REPO}/hparams/TRAINING/MEND/qwen2.5-7b.yaml"
    )
    hparams.model_name = MODEL
    hparams.tokenizer_name = MODEL
    hparams.model_parallel = mode == "multi"
    hparams.n_gpu = 3 if mode == "multi" else 1
    hparams.device = 0
    hparams.max_length = 64
    hparams.archive = None
    hparams.lr = 1e-6
    hparams.edit_lr = 1e-3
    hparams.alg = "MEND"
    hparams.train_base = False
    hparams.no_grad_layers = None
    hparams.silent = True
    hparams.accumulate_bs = 1
    hparams.inner_params = INNER_PARAMS
    return hparams


def make_batch(tok, device):
    prompts = [
        "Who founded the fictional city of Testville?",
        "What color is the imaginary mineral blorite?",
    ]
    targets = ["Professor Lina Vale", "bright teal"]
    enc = tok(
        [f"{p} {t}" for p, t in zip(prompts, targets)],
        return_tensors="pt",
        padding=True,
        truncation=True,
        max_length=64,
    )
    trg = tok(targets, return_tensors="pt", padding=True, truncation=True, max_length=64)
    labels = trg["input_ids"].clone()
    labels[labels == tok.pad_token_id] = -100
    return {
        "input_ids": enc["input_ids"].to(device),
        "attention_mask": enc["attention_mask"].to(device),
        "labels": labels.to(device),
    }


def grad_stats(params):
    grads = [p.grad for p in params if p.requires_grad]
    finite = [g for g in grads if g is not None and torch.isfinite(g).all()]
    nonzero = [g for g in finite if g.detach().float().abs().sum().item() > 0]
    return len(grads) - len(finite), len(nonzero), len(grads)


def run_case(mode):
    sys.path.insert(0, REPO)
    from easyeditor.trainer.algs.MEND import MEND, _model_forward
    from easyeditor.trainer.models import get_model
    from easyeditor.trainer.utils import _logits, safe_backward

    hparams = configure(mode)
    tok = AutoTokenizer.from_pretrained(MODEL, trust_remote_code=True)
    tok.pad_token = tok.pad_token or tok.eos_token

    base = get_model(hparams)
    base.bfloat16()
    mend = MEND(base, hparams, lambda: None)
    if hparams.model_parallel:
        hparams.device = mend.model.device
    else:
        mend.to(hparams.device)

    batch = make_batch(tok, hparams.device)
    kwargs = {k: v for k, v in batch.items() if k != "labels"}

    with torch.no_grad():
        base_before = _logits(_model_forward(mend.model, mend.config, **kwargs)).float().cpu()

    edited, _ = mend.edit(batch)
    edited_logits = edited(**batch)
    loss = mend.edit_loss_fn(mend.config, edited_logits, batch["labels"])["nll"]
    safe_backward(loss, mend.outer_parameters(), hparams.accumulate_bs, allow_unused=True)

    with torch.no_grad():
        base_after = _logits(_model_forward(mend.model, mend.config, **kwargs)).float().cpu()

    edit_diff = (edited_logits.detach().float().cpu() - base_before).abs().max().item()
    restore_diff = (base_after - base_before).abs().max().item()
    grad_none, grad_nonzero, grad_total = grad_stats(mend.outer_parameters())
    devices = sorted(
        {str(v) for v in getattr(mend.model, "hf_device_map", {}).values() if isinstance(v, int)}
    )

    ok = edit_diff > 0 and restore_diff == 0 and grad_none == 0 and grad_nonzero > 0
    print(
        f"{mode.upper()} visible={os.environ.get('CUDA_VISIBLE_DEVICES')} "
        f"devices={devices or ['0']} edit_diff={edit_diff:.6f} "
        f"restore_diff={restore_diff:.6f} grads={grad_nonzero}/{grad_total} "
        f"status={'OK' if ok else 'FAIL'}"
    )
    return 0 if ok else 1


def launch():
    cases = [("single", "1"), ("multi", "1,4,5")]
    for mode, visible in cases:
        env = os.environ.copy()
        env["CUDA_VISIBLE_DEVICES"] = visible
        cmd = [PYTHON, __file__, "--case", mode]
        result = subprocess.run(cmd, cwd=REPO, env=env, check=False)
        if result.returncode:
            return result.returncode
    return 0


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--case", choices=["single", "multi"])
    args = parser.parse_args()
    raise SystemExit(run_case(args.case) if args.case else launch())
